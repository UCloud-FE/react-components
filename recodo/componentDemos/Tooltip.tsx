import React from 'react';
import { Tooltip, Box, Button } from '@ucloud-fe/react-components';

const Demo = () => {
    return (
        <Box spacing="md">
            {['light', 'dark'].map(theme => (
                <Tooltip theme={theme} popup="提示信息" key={theme}>
                    <Button>按钮</Button>
                </Tooltip>
            ))}
        </Box>
    );
};

export default React.memo(Demo);
