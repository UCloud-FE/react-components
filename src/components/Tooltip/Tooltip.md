### 说明

*   Tooltip 基于 Popover，增加了箭头样式，支持所有 Popover 的 props
*   主要用于包裹一些展示性的提示

### 演示

*   演示

```js {"codepath": "tooltip.jsx"}
```

*   theme

```js {"codepath": "theme.jsx"}
```

*   arrow

```js {"codepath": "arrow.jsx"}
```

*   placement

```js {"codepath": "placement.jsx"}
```
