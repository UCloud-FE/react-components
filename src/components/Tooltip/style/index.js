import styled, { css, injectGlobal } from 'styled-components';

import { prefixCls as popoverPrefixCls } from 'src/components/Popover/style';
import addDefaultThemeProps from 'src/components/ThemeProvider/addDefaultThemeProps';
import config from 'config';

const { prefixCls: _prefixCls } = config;
export const prefixCls = _prefixCls + '-tooltip';

/* stylelint-disable property-no-unknown,no-duplicate-selectors */
const arrowWidth = '6px';
const borderWidth = '1px';

export const tooltipPopupClassName = prefixCls + '-popup';

export const ContentWrap = styled.div(({ theme: { designTokens: DT }, themeType }) => {
    let map = {
        light: {
            text: DT.T_COLOR_TEXT_DEFAULT_DARK,
            border: DT.T_POPOVER_COLOR_LINE_LIGHT,
            background: DT.T_POPOVER_COLOR_BG_LIGHT
        },
        dark: {
            text: DT.T_COLOR_TEXT_WHITE,
            border: DT.T_POPOVER_COLOR_BG_DARK,
            background: DT.T_POPOVER_COLOR_BG_DARK
        }
    };
    map = map[themeType];
    return css`
        padding: 8px 10px;
        text-align: left;
        text-decoration: none;
        border-radius: 3px;

        color: ${map.text};
        border: ${DT.T_LINE_WIDTH_BASE} solid ${map.border};
        background: ${map.background};
    `;
});

const arrowMixin = css`
    display: inline-block;
    position: absolute;
    width: 0;
    height: 0;
    border-width: 0;
    border-color: transparent;
    border-style: solid;
`;

export const Arrow = styled.span`
    ${arrowMixin};
`;
export const ArrowInner = styled.span`
    ${arrowMixin};
`;

injectGlobal`
    .${tooltipPopupClassName} {
        &.${popoverPrefixCls}-placement-bottom,
        &.${popoverPrefixCls}-placement-bottomLeft,
        &.${popoverPrefixCls}-placement-bottomRight {
            ${Arrow}, ${ArrowInner} {
                margin-left: -${arrowWidth};
                border-width: 0 ${arrowWidth} ${arrowWidth} ${arrowWidth};
                border-top-color: transparent;
                border-left-color: transparent;
                border-right-color: transparent;
            }
            ${Arrow} {
                top: -${arrowWidth};
            }
            ${ArrowInner} {
                top: ${borderWidth};
            }
        }
        &.${popoverPrefixCls}-placement-top,
        &.${popoverPrefixCls}-placement-topLeft,
        &.${popoverPrefixCls}-placement-topRight {
            ${Arrow}, ${ArrowInner} {
                margin-left: -${arrowWidth};
                border-width: ${arrowWidth} ${arrowWidth} 0 ${arrowWidth};
                border-bottom-color: transparent;
                border-left-color: transparent;
                border-right-color: transparent;
            }
            ${Arrow} {
                bottom: -${arrowWidth};
            }
            ${ArrowInner} {
                bottom: ${borderWidth};
            }
        }
        &.${popoverPrefixCls}-placement-left,
        &.${popoverPrefixCls}-placement-leftTop,
        &.${popoverPrefixCls}-placement-leftBottom {
            ${Arrow}, ${ArrowInner} {
                margin-top: -${arrowWidth};
                border-width: ${arrowWidth} 0 ${arrowWidth} ${arrowWidth};
                border-top-color: transparent;
                border-bottom-color: transparent;
                border-right-color: transparent;
            }
            ${Arrow} {
                right: -${arrowWidth};
            }
            ${ArrowInner} {
                right: ${borderWidth};
            }
        }
        &.${popoverPrefixCls}-placement-right,
        &.${popoverPrefixCls}-placement-rightTop,
        &.${popoverPrefixCls}-placement-rightBottom {
            ${Arrow}, ${ArrowInner} {
                margin-top: -${arrowWidth};
                border-width: ${arrowWidth} ${arrowWidth} ${arrowWidth} 0;
                border-top-color: transparent;
                border-bottom-color: transparent;
                border-left-color: transparent;
            }
            ${Arrow} {
                left: -${arrowWidth};
            }
            ${ArrowInner} {
                left: ${borderWidth};
            }
        }
        &.${popoverPrefixCls}-placement-bottomLeft,
        &.${popoverPrefixCls}-placement-topLeft {
            ${Arrow} {
                left: 15%;
            }
        }
        &.${popoverPrefixCls}-placement-bottomRight,
        &.${popoverPrefixCls}-placement-topRight {
            ${Arrow} {
                right: 15%;
            }
        }
        &.${popoverPrefixCls}-placement-leftTop,
        &.${popoverPrefixCls}-placement-rightTop {
            ${Arrow} {
                top: 15%;
                margin-top: 0;
            }
        }
        &.${popoverPrefixCls}-placement-leftBottom,
        &.${popoverPrefixCls}-placement-rightBottom {
            ${Arrow} {
                bottom: 15%;
            }
        }
        &.${popoverPrefixCls}-placement-top,
        &.${popoverPrefixCls}-placement-bottom {
            ${Arrow} {
                left: 50%;
            }
        }
        &.${popoverPrefixCls}-placement-left,
        &.${popoverPrefixCls}-placement-right {
            ${Arrow} {
                top: 50%;
            }
        }
    }
`;

export const TooltipWrap = styled.div(({ theme: { designTokens: DT }, themeType }) => {
    let map = {
        light: {
            border: DT.T_POPOVER_COLOR_LINE_LIGHT,
            background: DT.T_POPOVER_COLOR_BG_LIGHT
        },
        dark: {
            border: DT.T_POPOVER_COLOR_BG_DARK,
            background: DT.T_POPOVER_COLOR_BG_DARK
        }
    };
    map = map[themeType];
    return css`
        ${Arrow} {
            border-color: ${map.border};
        }
        ${ArrowInner} {
            border-color: ${map.background};
        }
    `;
});
addDefaultThemeProps(ContentWrap, TooltipWrap);
