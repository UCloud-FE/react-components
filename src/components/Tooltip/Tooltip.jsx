import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { ThemeProvider } from 'styled-components';

import Popover from 'components/Popover';
import { getPlacements } from 'components/Popover/placements';

import { TooltipWrap, themeMap, Arrow, ArrowInner, ContentWrap } from './style';

const { Animation, Trigger, Placement } = Popover;
const placements = getPlacements(7);
const Theme = ['light', 'dark'];

class Tooltip extends Component {
    static propTypes = {
        /** @ignore */
        popup: PropTypes.node.isRequired,
        /** @ignore */
        placement: PropTypes.oneOf(Placement),
        /** 主题风格 */
        theme: PropTypes.oneOfType([PropTypes.oneOf(Theme), PropTypes.object])
    };
    static defaultProps = {
        theme: 'light',
        placement: Placement[0]
    };
    renderPopup() {
        const { popup, placement, theme } = this.props;
        return (
            <ThemeProvider theme={_.isString(theme) ? themeMap[theme] : theme}>
                <TooltipWrap placement={placement}>
                    <Arrow>
                        <ArrowInner />
                    </Arrow>
                    <ContentWrap>{popup}</ContentWrap>
                </TooltipWrap>
            </ThemeProvider>
        );
    }
    render() {
        // eslint-disable-next-line no-unused-vars
        const { popup, ...rest } = this.props;
        return (
            <Popover
                getPopupContainer={triggerNode => triggerNode.parentNode}
                {...rest}
                builtinPlacements={placements}
                popup={this.renderPopup()}
            />
        );
    }
}

Object.assign(Tooltip, {
    Animation,
    Trigger,
    Placement,
    Theme
});

export default Tooltip;
