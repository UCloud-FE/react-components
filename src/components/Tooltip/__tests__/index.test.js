import React from 'react';
import { mount } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';

import Tooltip from 'src/components/Tooltip';

jest.unmock('rc-trigger');
describe('Tooltip', () => {
    test('tooltip', () => {
        const Demo = props => (
            <div>
                <Tooltip popup={<div>tooltip</div>} {...props}>
                    <div>content</div>
                </Tooltip>
            </div>
        );
        const wrapper = mount(<Demo />);
        expect(renderToJson(wrapper.render())).toMatchSnapshot();
        wrapper.setProps({
            visible: true
        });
        expect(renderToJson(wrapper.render())).toMatchSnapshot();
    });
});
