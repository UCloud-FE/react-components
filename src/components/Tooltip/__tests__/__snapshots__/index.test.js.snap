// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip tooltip 1`] = `
<div>
  <div>
    content
  </div>
</div>
`;

exports[`Tooltip tooltip 2`] = `
<div>
  <div
    style=""
  >
    content
  </div>
  <div
    style="position: absolute; top: 0px; left: 0px; width: 100%;"
  >
    <div>
      <div
        class="uc-fe-popover style__PopoverWrap-cfqvNq iMAKfT uc-fe-popover-placement-topLeft uc-fe-popover-animation-fade-appear"
        style="left: -999px; top: -1006px;"
      >
        <div
          class="style__TooltipWrap-gUIDEH coEymD"
        >
          <span
            class="style__Arrow-bfHBbq jAjcsO"
          >
            <span
              class="style__ArrowInner-cpbAEO jbBSfh"
            />
          </span>
          <div
            class="style__ContentWrap-kpQkHV iLFlIU"
          >
            <div>
              tooltip
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
